### ISSUE TITLE
issue with warmup baseline

### ISSUE THREAD

**Issue by LTluttmann** [wouterkool/attention-learn-to-route#51]
Hi Wouter,

first, big thanks for making the code available! I have a comment regarding the WarmupBaseline. As to my understanding, it should simply return the evaluation results of the "normal" baseline (self.baseline) once the number of warmup epochs is exceeded. However, the alpha attribute takes on values larger than 1  for all subsequent epochs: https://github.com/wouterkool/attention-learn-to-route/blob/6dbad47a415a87b5048df8802a74081a193fe240/reinforce_baselines.py#L64-L69

which causes the respective if-else statement to fail: https://github.com/wouterkool/attention-learn-to-route/blob/6dbad47a415a87b5048df8802a74081a193fe240/reinforce_baselines.py#L53-L62

So, I would propose either a clipping of the alpha attribute to a maximum value of 1 in the ```epoch_callback``` function or changing the respective if-statement to ```if self.alpha >= 1```

**Comment by LTluttmann** [comment]
Hi Wouter, thanks for your reply! That is indeed weird. I just debugged the code again and when using the Greedy-rollout baseline (also verified that bl_warmup_epochs=1), self.alpha indeed takes on values larger than 1 once the number of warmup epochs is exceeded:

<img width="849" alt="alpha" src="https://github.com/wouterkool/attention-learn-to-route/assets/66203600/637d0020-53d0-441e-8163-0b8562678c9d">


So, I started two training runs, one with alpha bounded to 1 and another where everything was kept as is. It showed, that not clipping the alpha leads to strongly decreasing baseline values and - as a result - to a growing loss that is backpropagated. Here are the tensorboard plots showing the baseline value, the REINFORCE loss and the gradient norm respectively (**yellow lines show the run with the clipped alpha, the pink lines the status quo**). 

![image](https://github.com/wouterkool/attention-learn-to-route/assets/66203600/06266732-b113-4571-b8de-87a058d0c817)

![image](https://github.com/wouterkool/attention-learn-to-route/assets/66203600/f61d6f8e-297b-48d5-b3a6-dc2dd5b2ca6d)

![image](https://github.com/wouterkool/attention-learn-to-route/assets/66203600/77f76e18-642d-4a53-bb3d-30360dc848f1)

**Clipping the alpha fixes these issues as can be seen from the plots, leading to much better validation performance:**

![image](https://github.com/wouterkool/attention-learn-to-route/assets/66203600/d6ab20bb-ce9c-4b60-8cf5-e8b7ffe8ec72)


Here are the full log files: [results.zip](https://github.com/wouterkool/attention-learn-to-route/files/11611675/results.zip)

**Comment by LTluttmann** [comment]
Hi @wouterkool, did you already manage to have a look at this?

**Comment by wouterkool** [comment]
Thanks a lot. Unfortunately I have no time to test it myself, however it seems that the fix you propose is correct, and as you tested it, I have merged it. I still don't understand why it did not cause any problems previously though...
